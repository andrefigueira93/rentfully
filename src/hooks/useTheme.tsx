import { themeAtom } from "@/store/global-store";
import { useAtom } from "jotai";
import { createContext, useContext, useEffect } from "react";

// Theme types
type Theme = "dark" | "light" | "system";

// Define the shape of the props
type ThemeProviderProps = {
  children: React.ReactNode;
};

// Define the shape of the context
type ThemeProviderState = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
};

// Define the initial state of the context
const initialState: ThemeProviderState = {
  theme: "system",
  setTheme: () => null,
};

// Create a context with the default value of undefined
const ThemeProviderContext = createContext<ThemeProviderState>(initialState);

// Create a provider component that will expose the context
export function ThemeProvider({ children, ...props }: ThemeProviderProps) {
  const [theme, setTheme] = useAtom(themeAtom);

  // Set the theme on first load
  useEffect(() => {
    const root = window.document.documentElement;

    root.classList.remove("light", "dark");

    if (theme === "system") {
      const systemTheme = window.matchMedia("(prefers-color-scheme: dark)")
        .matches
        ? "dark"
        : "light";

      root.classList.add(systemTheme);
      return;
    }

    root.classList.add(theme);
  }, [theme]);

  const value = {
    theme,
    setTheme: (theme: Theme) => {
      setTheme(theme);
    },
  };

  return (
    <ThemeProviderContext.Provider {...props} value={value}>
      {children}
    </ThemeProviderContext.Provider>
  );
}

// Create a hook to use the ThemeContext
export const useTheme = () => {
  const context = useContext(ThemeProviderContext);

  if (context === undefined)
    throw new Error("useTheme must be used within a ThemeProvider");

  return context;
};
