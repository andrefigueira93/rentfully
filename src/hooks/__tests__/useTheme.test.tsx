import TestWrapper from "@/__tests__/wrapper";
import { renderHook } from "@testing-library/react-hooks";
import { useTheme } from "../useTheme";

describe("useTheme hook", () => {
  test('should set the theme to "system" by default', () => {
    const { result } = renderHook(() => useTheme(), {
      wrapper: TestWrapper,
    });

    expect(result.current.theme).toBe("system");
    expect(result).toMatchSnapshot();
  });

  test("should set the theme to the provided value", () => {
    const { result } = renderHook(() => useTheme(), {
      wrapper: TestWrapper,
    });

    const newTheme = "dark";

    result.current.setTheme(newTheme);

    expect(result.current.theme).toBe(newTheme);
    expect(result).toMatchSnapshot();
  });
});
