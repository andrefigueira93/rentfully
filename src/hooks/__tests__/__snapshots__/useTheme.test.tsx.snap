// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useTheme hook > should set the theme to "system" by default 1`] = `
{
  "all": [
    {
      "setTheme": [Function],
      "theme": "system",
    },
  ],
  "current": {
    "setTheme": [Function],
    "theme": "system",
  },
  "error": undefined,
}
`;

exports[`useTheme hook > should set the theme to the provided value 1`] = `
{
  "all": [
    {
      "setTheme": [Function],
      "theme": "system",
    },
    {
      "setTheme": [Function],
      "theme": "dark",
    },
  ],
  "current": {
    "setTheme": [Function],
    "theme": "dark",
  },
  "error": undefined,
}
`;
